sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -725,4 +727,33 @@ trait HiveTypeCoercion {
         e.withNewChildren(newC)
     }
   }
+
+  /**
+   * Validate the data type for expression, and casts types as required, the expression should be
+   * [[ExpressionConstraint]].
+   */
+  object DataTypeConstraintValidation extends Rule[LogicalPlan] {
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
+      // Skip nodes who's children have not been resolved yet.
+      case e if !e.childrenResolved => e
+
+      case e: ExpressionConstraint =>
+        val newChildren = e.children.zip(e.constraint).map { case (expr, constraint) =>
```

Comment:
I thought that also, but this is will be solved during the debugging time, and in a release version, we never run into the case `e.children.length != e.constraint.length`, doesn't it?